#' barplot in reactable
#'
#' @name bar_style
#' @param length   the length of bar
#' @param height   the height of bar
#' @param fill     the bar color
#' @param align    the alignment of bar
#' @param color    the font color
#' @param fweight  the font weight
bar_style <- function(length = 1,
                      height = "80%",
                      fill = "#00FFFF7F",
                      align = c("left", "right"),
                      color = "black",
                      fweight = "plain") {
    align <- match.arg(align)
    if (align == "left") {
        position <- paste0(length * 100, "%")
        image <- sprintf("linear-gradient(90deg, %1$s %2$s, transparent %2$s)", fill, position)
    } else {
        position <- paste0(100 - length * 100, "%")
        image <- sprintf("linear-gradient(90deg, transparent %1$s, %2$s %1$s)", position, fill)
    }
    list(
        backgroundImage = image,
        backgroundSize = paste("100%", height),
        backgroundRepeat = "no-repeat",
        backgroundPosition = "center",
        color = color,
        fontWeight = fweight
    )
}

#' barplot in reactable
#'
#' @name bar_chart_pos_neg
#' @param lable     the length of bar
#' @param value     the height of bar
#' @param max_value the bar color
#' @param height    the alignment of bar
#' @param pos_fill  the font color
#' @param neg_fill  the font weight
bar_chart_pos_neg <- function(label,
                              value,
                              max_value = 1,
                              height = "1rem",
                              pos_fill = t_col("tomato", 0.8),
                              neg_fill = t_col("yellowgreen", 0.8)) {
    neg_chart <- div(style = list(flex = "1 1 0"))
    pos_chart <- div(style = list(flex = "1 1 0"))
    width <- paste0(abs(value / max_value) * 100, "%")

    if (value < 0) {
        bar <- div(style = list(marginLeft = "0.5rem", background = neg_fill, width = width, height = height))
        chart <- div(style = list(display = "flex", alignItems = "center", justifyContent = "flex-end"),
                     label,
                     bar)
        neg_chart <- tagAppendChild(neg_chart, chart)
    } else {
        bar <- div(style = list(marginRight = "0.5rem", background = pos_fill, width = width, height = height))
        chart <- div(style = list(display = "flex", alignItems = "center"), bar, label)
        pos_chart <- tagAppendChild(pos_chart, chart)
    }

    div(style = list(display = "flex"), neg_chart, pos_chart)
}

#' create QC reports
#'
#' @export
#' @name create_qc_reports
#' @param samplesheet    the path of sample sheet file
#' @param qc_type         screen or plasmid
#' @param qc_dir          the directory of QC plots and outs
create_qc_reports <- function(samplesheet = NULL,
                              qc_type = c("plasmid", "screen"),
                              qc_dir = NULL) {
        #----------#
        # checking #
        #----------#
        if (is.null(samplesheet)) {
            stop(paste0("====> Error: please provide the path of sample sheet file!"))
        }

        if (is.null(qc_dir)) {
            stop(paste0("====> Error: qc_dir is not provided, no output directory."))
        }

        if (!file.exists(paste0(qc_dir, "/sample_qc_cutoffs.tsv"))) {
            stop(paste0("====> Error: sample_qc_cutoffs.tsv is not in ", qc_dir, ". Please use qcout_samqc_cutoffs to create it."))
        }

        qc_type <- match.arg(qc_type)

        #------------------#
        # creating reports #
        #------------------#
        package_version <- paste0("MAVEQC", "-v", packageVersion("MAVEQC"))
        report_path <- paste0(qc_dir, "/", "MAVEQC_report.Rmd")
        sink(report_path)

        cat("---", "\n", sep = "")
        cat("title: \"MAVE QC Report\"", "\n", sep = "")
        cat("author: \"", package_version, "\"", "\n", sep = "")
        cat("date: \"`r format(Sys.time(), '%d %B %Y -- %A -- %X')`\"", "\n", sep = "")
        cat("output:", "\n", sep = "")
        cat("    html_document:", "\n", sep = "")
        cat("        toc: true", "\n", sep = "")
        cat("        toc_depth: 4", "\n", sep = "")
        cat("        theme: united", "\n", sep = "")
        cat("        highlight: tango", "\n", sep = "")
        cat("---", "\n", sep = "")
        cat("\n", sep = "")

        cat("```{r setup, include = FALSE}", "\n", sep = "")
        cat("knitr::opts_chunk$set(echo = TRUE, fig.align = \"center\")", "\n", sep = "")
        cat("library(reactable)", "\n", sep = "")
        cat("outdir <- \"", qc_dir, "\"", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("\n", sep = "")

        cat("```{js, echo = FALSE}", "\n", sep = "")
        cat("function formatNumber(num, precision = 1) {", "\n", sep = "")
        cat("    const map = [", "\n", sep = "")
        cat("        { suffix: 'T', threshold: 1e12 },", "\n", sep = "")
        cat("        { suffix: 'B', threshold: 1e9 },", "\n", sep = "")
        cat("        { suffix: 'M', threshold: 1e6 },", "\n", sep = "")
        cat("        { suffix: 'K', threshold: 1e3 },", "\n", sep = "")
        cat("        { suffix: '', threshold: 1 },", "\n", sep = "")
        cat("    ];", "\n", sep = "")
        cat("    const found = map.find((x) => Math.abs(num) >= x.threshold);", "\n", sep = "")
        cat("    if (found) {", "\n", sep = "")
        cat("        const formatted = (num / found.threshold).toFixed(precision) + found.suffix;", "\n", sep = "")
        cat("        return formatted;", "\n", sep = "")
        cat("    }", "\n", sep = "")
        cat("    return num;", "\n", sep = "")
        cat("}", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("function rangeMore(column, state) {", "\n", sep = "")
        cat("    let min = Infinity", "\n", sep = "")
        cat("    let max = 0", "\n", sep = "")
        cat("    state.data.forEach(function(row) {", "\n", sep = "")
        cat("        const value = row[column.id]", "\n", sep = "")
        cat("        if (value < min) {", "\n", sep = "")
        cat("            min = Math.floor(value)", "\n", sep = "")
        cat("        }", "\n", sep = "")
        cat("        if (value > max) {", "\n", sep = "")
        cat("            max = Math.ceil(value)", "\n", sep = "")
        cat("        }", "\n", sep = "")
        cat("    })", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("    const filterValue = column.filterValue || min", "\n", sep = "")
        cat("    const input = React.createElement('input', {", "\n", sep = "")
        cat("        type: 'range',", "\n", sep = "")
        cat("        value: filterValue,", "\n", sep = "")
        cat("        min: min,", "\n", sep = "")
        cat("        max: max,", "\n", sep = "")
        cat("        onChange: function(event) {", "\n", sep = "")
        cat("            column.setFilter(event.target.value || undefined)", "\n", sep = "")
        cat("        },", "\n", sep = "")
        cat("        style: { width: '100%', marginRight: '8px' },", "\n", sep = "")
        cat("        'aria-label': 'Filter ' + column.name", "\n", sep = "")
        cat("    })", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("    return React.createElement(", "\n", sep = "")
        cat("        'div',", "\n", sep = "")
        cat("        { style: { display: 'flex', alignItems: 'center', height: '100%' } },", "\n", sep = "")
        cat("        [input, formatNumber(filterValue)]", "\n", sep = "")
        cat("    )", "\n", sep = "")
        cat("}", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("function filterMinValue(rows, columnId, filterValue) {", "\n", sep = "")
        cat("    return rows.filter(function(row) {", "\n", sep = "")
        cat("        return row.values[columnId] >= filterValue", "\n", sep = "")
        cat("    })", "\n", sep = "")
        cat("}", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("function rangeLess(column, state) {", "\n", sep = "")
        cat("    let min = Infinity", "\n", sep = "")
        cat("    let max = 0", "\n", sep = "")
        cat("    state.data.forEach(function(row) {", "\n", sep = "")
        cat("        const value = row[column.id]", "\n", sep = "")
        cat("        if (value < min) {", "\n", sep = "")
        cat("            min = Math.floor(value)", "\n", sep = "")
        cat("        }", "\n", sep = "")
        cat("        if (value > max) {", "\n", sep = "")
        cat("            max = Math.ceil(value)", "\n", sep = "")
        cat("        }", "\n", sep = "")
        cat("    })", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("    const filterValue = column.filterValue || max", "\n", sep = "")
        cat("    const input = React.createElement('input', {", "\n", sep = "")
        cat("        type: 'range',", "\n", sep = "")
        cat("        value: filterValue,", "\n", sep = "")
        cat("        min: min,", "\n", sep = "")
        cat("        max: max,", "\n", sep = "")
        cat("        onChange: function(event) {", "\n", sep = "")
        cat("            column.setFilter(event.target.value || undefined)", "\n", sep = "")
        cat("        },", "\n", sep = "")
        cat("        style: { width: '100%', marginRight: '8px' },", "\n", sep = "")
        cat("        'aria-label': 'Filter ' + column.name", "\n", sep = "")
        cat("    })", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("    return React.createElement(", "\n", sep = "")
        cat("        'div',", "\n", sep = "")
        cat("        { style: { display: 'flex', alignItems: 'center', height: '100%' } },", "\n", sep = "")
        cat("        [input, formatNumber(filterValue)]", "\n", sep = "")
        cat("    )", "\n", sep = "")
        cat("}", "\n", sep = "")
        cat("", "\n", sep = "")
        cat("function filterMaxValue(rows, columnId, filterValue) {", "\n", sep = "")
        cat("    return rows.filter(function(row) {", "\n", sep = "")
        cat("        return row.values[columnId] <= filterValue", "\n", sep = "")
        cat("    })", "\n", sep = "")
        cat("}", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("\n", sep = "")

        cat("---", "\n", sep = "")
        cat("\n", sep = "")

        cat("## 1. Introduction", "\n", sep = "")
        cat("MAVEQC is a flexible R-package that provides QC analysis of Saturation Genome Editing (SGE) experimental data. ",
            "Available under GPL 3.0 from https://github.com/wtsi-hgi/MAVEQC", "\n", sep = "")
        cat("\n", sep = "")

        cat("---", "\n", sep = "")
        cat("\n", sep = "")

        if (qc_type == "screen") {
            cat("## 2. Screen QC", "\n", sep = "")
        } else {
            cat("## 2. Plasmid QC", "\n", sep = "")
        }
        cat("Displays QC plots and statistics for all samples for QC.", "\n", sep = "")
        cat("\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("samqc_cutoffs <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_cutoffs.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("### 2.1. Sample Sheet", "\n", sep = "")
        cat("\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", samplesheet, "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")))", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_meta.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")))", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("### 2.2. Run Sample QC", "\n", sep = "")
        cat("\n", sep = "")
        cat("#### 2.2.1. Read Length Distribution", "\n", sep = "")
        cat("Displays the percentage of reads for each sample, based on 50 nucleotide increments, using the total number of raw reads.", "\n", sep = "")
        cat("<p style=\"color:red\">Note: expected read length is 300.</p>", "\n", sep = "")
        cat("<p style=\"color:red\">Note: the lengths of primers are deducted from the read length based on the sample sheet information. ",
            "(see 2.1. Sample Sheet: quants_append_start and quants_append_end)</p>", "\n", sep = "")
        cat("\n", sep = "")

        cat("```{r, echo = FALSE, out.height = \"75%\", out.width = \"75%\"}", "\n", sep = "")
        cat("knitr::include_graphics(paste0(outdir, \"/sample_qc_read_length.png\"), rel_path = FALSE)", "\n", sep = "")
        cat("```", "\n", sep = "")

        cat("**Pass criterion:** more than 90% of reads are longer than 200 nucleotides", "\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_read_length.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"% 0 ~ 50\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                               style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"% 50 ~ 100\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                 style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"% 100 ~ 150\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"% 150 ~ 200\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"% 200 ~ 250\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"% 250 ~ 300\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"Total Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Pass Threshold (%)\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Pass\" = colDef(minWidth = 50, cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })),", "\n", sep = "")
        cat("          rowStyle = function(index) { if (!(df[index, \"Pass\"])) { list(background = t_col(\"tomato\", 0.2)) }})", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("#### 2.2.2. Missing Variants", "\n", sep = "")
        cat("Stats of missing variants in the library", "\n", sep = "")
        cat("\n", sep = "")
        cat("**Pass criterion:** less than 1% of expected variants are missing", "\n", sep = "")
        cat("\n", sep = "")

        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_missing.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"Library Sequences\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Missing Library Sequences\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"% Missing Library Sequences\" = colDef(minWidth = 100, style = function(value) { ", "\n", sep = "")
        cat("                                                                                          if (value > samqc_cutoffs$per_missing_variants * 100) {", "\n", sep = "")
        cat("                                                                                              color <- \"red\"", "\n", sep = "")
        cat("                                                                                              fweight <- \"bold\"", "\n", sep = "")
        cat("                                                                                          } else {", "\n", sep = "")
        cat("                                                                                              color <- \"forestgreen\"", "\n", sep = "")
        cat("                                                                                              fweight <- \"bold\" }", "\n", sep = "")
        cat("                                                                                          list(color = color, fontWeight = fweight)}),", "\n", sep = "")
        cat("                         \"Pass Threshold (%)\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Pass\" = colDef(minWidth = 50, cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })),", "\n", sep = "")
        cat("          rowStyle = function(index) { if (!(df[index, \"Pass\"])) { list(background = t_col(\"tomato\", 0.2)) }})", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("Records of missing variants in the library", "\n", sep = "")
        cat("<p style=\"color:red\">Note: Unique indicates that a template sequence occurs only once in the VaLiAnT meta file. (1: Unique, 0: Not Unique) ",
            "This is important as a template sequence can occur more than once depending on the mutation types applied in VaLiAnT.</p>", "\n", sep = "")
        cat("<p style=\"color:red\">Note: Table below shows all the missing variants in all the samples, so the variants may occur multiple times.</p>", "\n", sep = "")

        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/missing_variants_in_library.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("df <- df[, c(\"name\", \"length\", \"count\", \"unique\", \"sample\")]", "\n", sep = "")
        cat("colnames(df) <- capital_names(colnames(df))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 50, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Name\" = colDef(minWidth = 300),", "\n", sep = "")
        cat("                         \"Sample\" = colDef(minWidth = 150)))", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("#### 2.2.3. Total Reads (Counts)", "\n", sep = "")
        cat("Displays the total number of reads per sample. ",
            "Filtering based on 1-dimensional Kmean clustering that excludes unique sequences with low read counts.", "\n", sep = "")
        cat("\n", sep = "")
        cat("* **Accepted reads:** Total read count for all unique sequences with sufficient reads based 1D Kmean clustering.", "\n", sep = "")
        cat("* **Excluded reads:** Total read count for all unique sequences with insufficient reads based 1D Kmean clustering.", "\n", sep = "")
        cat("\n", sep = "")

        cat("```{r, echo = FALSE, out.height = \"75%\", out.width = \"75%\"}", "\n", sep = "")
        cat("knitr::include_graphics(paste0(outdir, \"/sample_qc_stats_total.png\"), rel_path = FALSE)", "\n", sep = "")
        cat("```", "\n", sep = "")

        cat("Total Reads: the total number of raw reads", "\n", sep = "")
        cat("\n", sep = "")
        cat("**Pass criterion:** more than 1,000,000 total reads", "\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_total.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"Accepted Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                     format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"% Accepted Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                       style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"Excluded Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                     format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"% Excluded Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                       style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"Total Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  format = colFormat(separators = TRUE),", "\n", sep = "")
        cat("                                                  style = function(value) { ", "\n", sep = "")
        cat("                                                              if (value < samqc_cutoffs$num_total_reads) {", "\n", sep = "")
        cat("                                                                  color <- \"red\"", "\n", sep = "")
        cat("                                                                  fweight <- \"bold\"", "\n", sep = "")
        cat("                                                              } else {", "\n", sep = "")
        cat("                                                                  color <- \"forestgreen\"", "\n", sep = "")
        cat("                                                                  fweight <- \"bold\" }", "\n", sep = "")
        cat("                                                              list(color = color, fontWeight = fweight)}),", "\n", sep = "")
        cat("                         \"Pass Threshold\" = colDef(minWidth = 100,format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Pass\" = colDef(minWidth = 50, cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })),", "\n", sep = "")
        cat("          rowStyle = function(index) { if (!(df[index, \"Pass\"])) { list(background = t_col(\"tomato\", 0.2)) }})", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("#### 2.2.4. Accepted Reads (Percentage)", "\n", sep = "")
        cat("Displays the percentage of library reads vs non-library reads (ie. Reference, PAM and Unmapped) for Accepted Reads (see 2.2.3 explanation).", "\n", sep = "")
        cat("\n", sep = "")
        cat("* **Library Reads:** Percentage reads mapping to template oligo sequences, including intended variants.", "\n", sep = "")
        cat("* **Reference Reads:** Percentage reads mapping to Reference.", "\n", sep = "")
        cat("* **PAM Reads:** Percentage reads mapping to PAM/Protospacer Protection Edits (PPEs), without intended variant.", "\n", sep = "")
        cat("* **Unmapped Reads:** Percentage of Unmapped Reads (not mapped to library sequences, PAM sequence, and reference sequence).", "\n", sep = "")
        cat("* **Library Coverage:** Mean read count per template oligo sequence.", "\n", sep = "")
        cat("\n", sep = "")

        cat("\n", sep = "")
        df <- as.data.frame(read.table(paste0(qc_dir, "/sample_qc_stats_accepted.tsv"), header = TRUE, sep = "\t", check.names = FALSE))
        samples_ref0 <- df[df[, 4] == 0, 2]
        samples_pam0 <- df[df[, 5] == 0, 2]
        if (length(samples_ref0) != 0) {
            cat("<p style=\"color:red; font-weight: bold\">Warning: found samples with 0 counts of resequence sequence! Please check table below.</p>", "\n", sep = "")
        }
        if (length(samples_pam0) != 0) {
            cat("<p style=\"color:red; font-weight: bold\">Warning: found samples with 0 counts of pam sequence! Please check table below.</p>", "\n", sep = "")
        }
        cat("\n", sep = "")

        cat("```{r, echo = FALSE, out.height = \"75%\", out.width = \"75%\"}", "\n", sep = "")
        cat("knitr::include_graphics(paste0(outdir, \"/sample_qc_stats_accepted.png\"), rel_path = FALSE)", "\n", sep = "")
        cat("```", "\n", sep = "")

        cat("**Pass criterion:** more than 40% of accepted reads are library reads", "\n", sep = "")
        cat("\n", sep = "")
        cat("<p style=\"color:red\">Note: Accepted reads are the filtered reads based on 2.2.3</p>", "\n", sep = "")

        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_accepted.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("qc_type_exists <- \"screen\" %in% df$`QC_Type`", "\n", sep = "")
        cat("qa_pass_sr <- function(lr_value) {ifelse(lr_value < 30, \"\\u274c\", ifelse(lr_value >= 30 & lr_value < 40, \"\\u2705\\u2757\", \"\\u2705\"))}", "\n", sep = "")
        cat("qa_pass_pl <- function(pass_value) { ifelse(pass_value, \"\\u2705\", \"\\u274c\") }", "\n", sep = "")
        cat("if (!qc_type_exists) df$Pass <- qa_pass_pl(df$Pass) else df <- df %>% mutate(\"Pass\" = qa_pass_sr(df$`% Library Reads`))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"% Library Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                      style = function(value) {", "\n", sep = "")
        cat("                                                                  if (value < samqc_cutoffs$per_library_reads * 100) {", "\n", sep = "")
        cat("                                                                      bar_style(length = value/100, color = \"red\", fweight = \"bold\")", "\n", sep = "")
        cat("                                                                  } else {", "\n", sep = "")
        cat("                                                                      bar_style(length = value/100, color = \"forestgreen\", fweight = \"bold\") }}),", "\n", sep = "")
        cat("                         \"% Reference Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                        style = function(value) {", "\n", sep = "")
        cat("                                                                    if (value == 0) {", "\n", sep = "")
        cat("                                                                        bar_style(length = value/100, color = \"red\", fweight = \"bold\")", "\n", sep = "")
        cat("                                                                    } else {", "\n", sep = "")
        cat("                                                                        bar_style(length = value/100)}}),", "\n", sep = "")
        cat("                         \"% PAM Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                  style = function(value) {", "\n", sep = "")
        cat("                                                                    if (value == 0) {", "\n", sep = "")
        cat("                                                                        bar_style(length = value/100, color = \"red\", fweight = \"bold\")", "\n", sep = "")
        cat("                                                                    } else {", "\n", sep = "")
        cat("                                                                        bar_style(length = value/100)}}),", "\n", sep = "")
        cat("                         \"% Unmapped Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                       style = function(value) {bar_style(length = value/100)}),", "\n", sep = "")
        cat("                         \"Pass Threshold (%)\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Pass\" = colDef(minWidth = 50, filterable = FALSE),", "\n", sep = "")
        cat("                         \"QC_Type\" = colDef(show = FALSE)))", "\n", sep = "")
        cat("```", "\n", sep = "")

        cat("\n", sep = "")

        cat("Defines the mean read count per template oligo sequence (dividing the total number of library reads by the total number of library sequences).", "\n", sep = "")
        cat("\n", sep = "")
        cat("**Pass criterion:** library coverage is more than 100 reads", "\n", sep = "")

        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_coverage.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"Total Library Reads\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                          format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Total Library Sequences\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                                     format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Library Coverage\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                       format = colFormat(separators = TRUE),", "\n", sep = "")
        cat("                                                       style = function(value) {", "\n", sep = "")
        cat("                                                                   if (value < samqc_cutoffs$library_cov) {", "\n", sep = "")
        cat("                                                                       color <- \"red\"", "\n", sep = "")
        cat("                                                                       fweight <- \"bold\"", "\n", sep = "")
        cat("                                                                   } else {", "\n", sep = "")
        cat("                                                                       color <- \"forestgreen\"", "\n", sep = "")
        cat("                                                                       fweight <- \"bold\" }", "\n", sep = "")
        cat("                                                                   list(color = color, fontWeight = fweight)}),", "\n", sep = "")
        cat("                         \"Median Coverage\" = colDef(filterMethod = JS(\"filterMinValue\"), filterInput = JS(\"rangeMore\"),", "\n", sep = "")
        cat("                                                      format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Pass Threshold\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Pass\" = colDef(minWidth = 50, cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })),", "\n", sep = "")
        cat("          rowStyle = function(index) { if (!(df[index, \"Pass\"])) { list(background = t_col(\"tomato\", 0.2)) }})", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("#### 2.2.5. Genomic Coverage", "\n", sep = "")
        cat("Distribution of variants across targeton region based on log2(count+1) values.", "\n", sep = "")
        if (qc_type == "screen") {
            cat("<p style=\"color:red\">Note: Does not show missing varaints (0 count in the libary).</p>", "\n", sep = "")
        }
        cat("\n", sep = "")

        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("knitr::include_graphics(paste0(outdir, \"/sample_qc_position_cov.dots.png\"), rel_path = FALSE)", "\n", sep = "")
        cat("```", "\n", sep = "")

        cat("Low Abundance cutoff: the green dashed line indicates the threshold which is used to determine if the variant is low abundance (less than 5 reads)", "\n", sep = "")
        cat("\n", sep = "")
        cat("**Pass criterion:** the percentage of low-abundance variants is lower than 30%", "\n", sep = "")
        cat("\n", sep = "")
        cat("% Low Abundance: the percentage of variants below the low abundance cutoff", "\n", sep = "")
        cat("\n", sep = "")

        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_pos_coverage.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("df_counts <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_pos_counts.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("df_counts <- log2(df_counts + 1)", "\n", sep = "")
        cat("df$data <- NA", "\n", sep = "")
        cat("for (i in 1:nrow(df)) {", "\n", sep = "")
        cat("    tmp_data <- df_counts[, df[i, ]$Sample]", "\n", sep = "")
        cat("    df[i, ]$data <- list(tmp_data[!is.na(tmp_data)])", "\n", sep = "")
        cat("}", "\n", sep = "")
        cat("df$boxplot <- NA", "\n", sep = "")
        cat("boxplot_min <- min(df_counts, na.rm = TRUE)", "\n", sep = "")
        cat("boxplot_max <- max(df_counts, na.rm = TRUE)", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df[, c(1:4, 14, 9, 5:8, 10:12)], highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"Chromosome\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Strand\" = colDef(minWidth = 60),", "\n", sep = "")
        cat("                         \"Genomic Start\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"Genomic End\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
        cat("                         \"boxplot\" = colDef(minWidth = 200, cell = function(value, index) {", "\n", sep = "")
        cat("                                                         if (length(df$data[[index]]) > 5) {", "\n", sep = "")
        cat("                                                             sparkline(df$data[[index]], type = \"box\", width = 180,", "\n", sep = "")
        cat("                                                                       chartRangeMin = boxplot_min, chartRangeMax = boxplot_max) }}),", "\n", sep = "")
        cat("                         \"% Low Abundance\" = colDef(minWidth = 100, style = function(value) {", "\n", sep = "")
        cat("                                                                  if (value > (1 - samqc_cutoffs$low_abundance_lib_per) * 100) {", "\n", sep = "")
        cat("                                                                      color <- \"red\"", "\n", sep = "")
        cat("                                                                      fweight <- \"bold\"", "\n", sep = "")
        cat("                                                                  } else {", "\n", sep = "")
        cat("                                                                      color <- \"forestgreen\"", "\n", sep = "")
        cat("                                                                      fweight <- \"bold\" }", "\n", sep = "")
        cat("                                                                  list(color = color, fontWeight = fweight)}),", "\n", sep = "")
        cat("                         \"Low Abundance cutoff\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Pass Threshold (%)\" = colDef(minWidth = 100),", "\n", sep = "")
        cat("                         \"Pass\" = colDef(minWidth = 50, cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })),", "\n", sep = "")
        cat("          rowStyle = function(index) { if (!(df[index, \"Pass\"])) { list(background = t_col(\"tomato\", 0.2)) }})", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        if (qc_type == "screen") {
            cat("#### 2.2.6. Genomic Position Percentage", "\n", sep = "")
            cat("Displays distribution of \"LOF\" (loss-of-function) vs all \"Other\" variants across the targeton region, ",
                "based on read percentages for reference timepoint. ",
                "Requires concordant distribution of LOF and Other variants.", "\n", sep = "")
            cat("<p style=\"color:red\">Note: Does not show missing varaints (0 count in the libary).</p>", "\n", sep = "")
            cat("\n", sep = "")

            cat("```{r, echo = FALSE, out.height = \"65%\", out.width = \"65%\"}", "\n", sep = "")
            cat("knitr::include_graphics(paste0(outdir, \"/sample_qc_position_anno.lof_dots.png\"), rel_path = FALSE)", "\n", sep = "")
            cat("```", "\n", sep = "")
            cat("```{r, echo = FALSE}", "\n", sep = "")
            cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_stats_pos_percentage.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
            cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
            cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
            cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
            cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
            cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
            cat("                         \"Chromosome\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"Strand\" = colDef(minWidth = 60),", "\n", sep = "")
            cat("                         \"Genomic Start\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
            cat("                         \"Genomic End\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
            cat("                         \"% Low Abundance (LOF)\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"% Low Abundance (Others)\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"% Low Abundance (ALL)\" = colDef(minWidth = 100, style = function(value) {", "\n", sep = "")
            cat("                                                                        if (value > (1 - samqc_cutoffs$low_abundance_lib_per) * 100) {", "\n", sep = "")
            cat("                                                                            color <- \"red\"", "\n", sep = "")
            cat("                                                                            fweight <- \"bold\"", "\n", sep = "")
            cat("                                                                        } else {", "\n", sep = "")
            cat("                                                                            color <- \"forestgreen\"", "\n", sep = "")
            cat("                                                                            fweight <- \"bold\" }", "\n", sep = "")
            cat("                                                                        list(color = color, fontWeight = fweight)}),", "\n", sep = "")
            cat("                         \"% Low Abundance cutoff\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"Pass Threshold\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"Pass\" = colDef(minWidth = 50, cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })),", "\n", sep = "")
            cat("          rowStyle = function(index) { if (!(df[index, \"Pass\"])) { list(background = t_col(\"tomato\", 0.2)) }})", "\n", sep = "")
            cat("```", "\n", sep = "")
            cat("<br>", "\n", sep = "")
            cat("\n", sep = "")

            cat("### 2.3. Run Experiment QC", "\n", sep = "")
            cat("\n", sep = "")
            cat("#### 2.3.1. Sample Correlations", "\n", sep = "")
            cat("\n", sep = "")
            cat("```{r, echo = FALSE, out.height = \"50%\", out.width = \"50%\"}", "\n", sep = "")
            cat("knitr::include_graphics(paste0(outdir, \"/experiment_qc_samples_tree.png\"), rel_path = FALSE)", "\n", sep = "")
            cat("knitr::include_graphics(paste0(outdir, \"/experiment_qc_samples_corr.png\"), rel_path = FALSE)", "\n", sep = "")
            cat("```", "\n", sep = "")
            cat("<br>", "\n", sep = "")
            cat("\n", sep = "")
            cat("```{r, echo = FALSE}", "\n", sep = "")
            cat("df <- as.data.frame(read.table(\"", qc_dir, "/experiment_qc_corr.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
            cat("df$Pass <- NULL", "\n", sep = "")
            cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
            cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
            cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
            cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
            cat("          columns = list(\"Cluster\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"Replicate\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"Condition\" = colDef(minWidth = 100),", "\n", sep = "")
            cat("                         \"Pass\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" })))", "\n", sep = "")
            cat("```", "\n", sep = "")
            cat("<br>", "\n", sep = "")
            cat("\n", sep = "")

            cat("#### 2.3.2. Sample PCA", "\n", sep = "")
            cat("\n", sep = "")
            cat("```{r, echo = FALSE, out.height = \"75%\", out.width = \"75%\"}", "\n", sep = "")
            cat("knitr::include_graphics(paste0(outdir, \"/experiment_qc_pca_samples.png\"), rel_path = FALSE)", "\n", sep = "")
            cat("```", "\n", sep = "")
            cat("<br>", "\n", sep = "")
            cat("\n", sep = "")

            cat("#### 2.3.3. Fold Change (by category)", "\n", sep = "")
            cat("\n", sep = "")
            figs <- list.files(path = qc_dir, pattern = "experiment_qc_deseq_fc.*.all_beeswarm.png", full.names = TRUE)
            figs <- mixedsort(figs)
            tsvs <- list.files(path = qc_dir, pattern = "experiment_qc_deseq_fc.*.all.tsv", full.names = TRUE)
            tsvs <- mixedsort(tsvs)
            sum_tsvs <- list.files(path = qc_dir, pattern = "experiment_qc_deseq_fc.*.all_sum.tsv", full.names = TRUE)
            sum_tsvs <- mixedsort(sum_tsvs)
            for (i in 1:length(figs)) {
                tmp_header <- strsplit(tail(strsplit(figs[i], "/", fixed = TRUE)[[1]], n = 1), ".", fixed = TRUE)[[1]][2]
                cat("**", tmp_header, "**", "\n", sep = "")

                cat("```{r, echo = FALSE, out.height = \"100%\", out.width = \"100%\"}", "\n", sep = "")
                cat("knitr::include_graphics(\"", figs[i], "\", rel_path = FALSE)", "\n", sep = "")
                cat("```", "\n", sep = "")
                cat("\n", sep = "")

                cat("```{r, echo = FALSE}", "\n", sep = "")
                cat("df <- as.data.frame(read.table(\"", tsvs[i], "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
                cat("df_sum <- as.data.frame(read.table(\"", sum_tsvs[i], "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
                cat("df_sum$data <- NA", "\n", sep = "")
                cat("for (i in 1:nrow(df_sum)) {", "\n", sep = "")
                cat("    df_sum[i, ]$data <- list(sort(df[df$consequence == df_sum[i, ]$consequence, ]$adj_log2FoldChange))", "\n", sep = "")
                cat("}", "\n", sep = "")
                cat("df_sum$boxplot <- NA", "\n", sep = "")
                cat("df_sum$barplot <- NA", "\n", sep = "")
                cat("min_row <- ifelse(nrow(df_sum) > 10, 10, nrow(df_sum))", "\n", sep = "")
                cat("boxplot_min <- min(df$adj_log2FoldChange, na.rm = TRUE)", "\n", sep = "")
                cat("boxplot_max <- max(df$adj_log2FoldChange, na.rm = TRUE)", "\n", sep = "")
                cat("reactable(df_sum[, c(1, 9, 10, 2:7)], highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
                cat("          filterable = FALSE, sortable = FALSE, minRows = min_row, defaultColDef = colDef(minWidth = 200, align = \"left\"),", "\n", sep = "")
                cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
                cat("          columns = list(\"consequence\" = colDef(minWidth = 250),", "\n", sep = "")
                cat("                         \"number of depleted\"  = colDef(minWidth = 100, style = function(value) { list(color = \"forestgreen\", fontWeight = \"bold\") }),", "\n", sep = "")
                cat("                         \"number of no impact\"  = colDef(minWidth = 100, style = function(value) { list(color = \"black\", fontWeight = \"bold\") }),", "\n", sep = "")
                cat("                         \"number of enriched\"  = colDef(minWidth = 100, style = function(value) { list(color = \"red\", fontWeight = \"bold\") }),", "\n", sep = "")
                cat("                         \"lfc range\" = colDef(minWidth = 100),", "\n", sep = "")
                cat("                         \"total number\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
                cat("                         \"total number of outside range\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
                cat("                         \"boxplot\" = colDef(cell = function(value, index) {", "\n", sep = "")
                cat("                                                         if (length(df_sum$data[[index]]) > 5) {", "\n", sep = "")
                cat("                                                             sparkline(df_sum$data[[index]], type = \"box\", width = 180,", "\n", sep = "")
                cat("                                                                       chartRangeMin = boxplot_min, chartRangeMax = boxplot_max) }}),", "\n", sep = "")
                cat("                         \"barplot\" = colDef(cell = function(value, index) {", "\n", sep = "")
                cat("                                                         if (length(df_sum$data[[index]]) > 5) {", "\n", sep = "")
                cat("                                                             sparkline(df_sum$data[[index]], type = \"tristate\", width = 180,", "\n", sep = "")
                cat("                                                                       posBarColor = \"red\", negBarColor = \"yellowgreen\") }}) ))", "\n", sep = "")
                cat("```", "\n", sep = "")
                cat("\n", sep = "")
            }
            cat("<br>", "\n", sep = "")
            cat("\n", sep = "")

            cat("#### 2.3.4. Fold Change (by position)", "\n", sep = "")
            cat("\n", sep = "")
            figs <- list.files(path = qc_dir, pattern = "experiment_qc_deseq_fc.*.all_position.png", full.names = TRUE)
            figs <- mixedsort(figs)
            tsvs <- list.files(path = qc_dir, pattern = "experiment_qc_deseq_fc.*.all.tsv", full.names = TRUE)
            tsvs <- mixedsort(tsvs)
            for (i in 1:length(figs)) {
                tmp_header <- strsplit(tail(strsplit(figs[i], "/", fixed = TRUE)[[1]], n = 1), ".", fixed = TRUE)[[1]][2]
                cat("**", tmp_header, "**", "\n", sep = "")

                cat("```{r, echo = FALSE, out.height = \"100%\", out.width = \"100%\"}", "\n", sep = "")
                cat("knitr::include_graphics(\"", figs[i], "\", rel_path = FALSE)", "\n", sep = "")
                cat("```", "\n", sep = "")
                cat("\n", sep = "")

                cat("```{r, echo = FALSE}", "\n", sep = "")
                cat("df <- as.data.frame(read.table(\"", tsvs[i], "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
                cat("df <- df[df$stat != \"no impact\", c(\"oligo_name\", \"consequence\", \"position\", \"adj_log2FoldChange\", \"adj_fdr\", \"stat\")]", "\n", sep = "")
                cat("max_log2fc <- ifelse(nrow(df) == 0, 0, max(abs(df$adj_log2FoldChange)))", "\n", sep = "")
                cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
                cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
                cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
                cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
                cat("          columns = list(\"oligo_name\" = colDef(minWidth = 300),", "\n", sep = "")
                cat("                         \"position\" = colDef(minWidth = 100, format = colFormat(separators = TRUE)),", "\n", sep = "")
                cat("                         \"adj_log2FoldChange\" = colDef(filterMethod = JS(\"filterMaxValue\"),", "\n", sep = "")
                cat("                                                     filterInput = JS(\"rangeLess\"),", "\n", sep = "")
                cat("                                                     format = colFormat(digits = 2),", "\n", sep = "")
                cat("                                                     cell = function(value) {", "\n", sep = "")
                cat("                                                                label <- round(value, 2)", "\n", sep = "")
                cat("                                                                bar_chart_pos_neg(label, value, max_value = max_log2fc)},", "\n", sep = "")
                cat("                                                     align = \"center\",", "\n", sep = "")
                cat("                                                     minWidth = 300),", "\n", sep = "")
                cat("                         \"adj_fdr\" = colDef(minWidth = 100, format = colFormat(digits = 3)),", "\n", sep = "")
                cat("                         \"stat\"  = colDef(minWidth = 100, style = function(value) {", "\n", sep = "")
                cat("                                                        if (value == \"enriched\") {", "\n", sep = "")
                cat("                                                            color <- \"red\"", "\n", sep = "")
                cat("                                                            fweight <- \"bold\"", "\n", sep = "")
                cat("                                                        } else {", "\n", sep = "")
                cat("                                                            color <- \"forestgreen\"", "\n", sep = "")
                cat("                                                            fweight <- \"bold\" }", "\n", sep = "")
                cat("                                                        list(color = color, fontWeight = fweight)})))", "\n", sep = "")
                cat("```", "\n", sep = "")
                cat("\n", sep = "")
            }
            cat("<br>", "\n", sep = "")
            cat("\n", sep = "")
        }

        cat("---", "\n", sep = "")
        cat("\n", sep = "")

        cat("## 3. QC Results", "\n", sep = "")
        cat("Summarising the final results, below are the cutoffs using for PASS/FAIL", "\n", sep = "")
        cat("\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(matrix(1:21, nrow = 7))", "\n", sep = "")
        cat("colnames(df) <- c(\"QC Tag\", \"Cutoff\", \"Description\")", "\n", sep = "")
        cat("df[, 1] <- c(\"Total Reads\",", "\n", sep = "")
        cat("             \"% Missing Variants\",", "\n", sep = "")
        cat("             \"Accepted Reads\",", "\n", sep = "")
        cat("             \"% Mapping Reads\",", "\n", sep = "")
        cat("             \"% Reference Reads\",", "\n", sep = "")
        cat("             \"% Library Reads\",", "\n", sep = "")
        cat("             \"Library Coverage\")", "\n", sep = "")
        cat("df[, 2] <- c(samqc_cutoffs$num_total_reads,", "\n", sep = "")
        cat("             samqc_cutoffs$per_missing_variants,", "\n", sep = "")
        cat("             samqc_cutoffs$num_accepted_reads,", "\n", sep = "")
        cat("             samqc_cutoffs$per_mapping_reads,", "\n", sep = "")
        cat("             samqc_cutoffs$per_ref_reads,", "\n", sep = "")
        cat("             samqc_cutoffs$per_library_reads,", "\n", sep = "")
        cat("             samqc_cutoffs$library_cov)", "\n", sep = "")
        cat("df[, 3] <- c(\"Sample must have more than [cutoff] total reads\",", "\n", sep = "")
        cat("             \"Missing varaints in the library must be less than [cutoff]\",", "\n", sep = "")
        cat("             \"Sample must have more than [cutoff] reads after the low count filtering\",", "\n", sep = "")
        cat("             \"Sample must have more than [cutoff] of reads aligned to the library including reference and PAM reads\",", "\n", sep = "")
        cat("             \"Sample must have less than [cutoff] of reads aligned to reference sequence\",", "\n", sep = "")
        cat("             \"Sample must have more than [cutoff] of reads aligned to the library\",", "\n", sep = "")
        cat("             \"Sample must have more than [cutoff] average coverage\")", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          minRows = min_row, defaultColDef = colDef(minWidth = 150, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"QC Tag\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"Description\" = colDef(minWidth = 800)))", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("### 3.1. Sample QC Results", "\n", sep = "")
        cat("\n", sep = "")
        cat("```{r, echo = FALSE}", "\n", sep = "")
        cat("df <- as.data.frame(read.table(\"", qc_dir, "/sample_qc_results.tsv", "\", header = TRUE, sep = \"\\t\", check.names = FALSE))", "\n", sep = "")
        cat("col_pal <- function(x) rgb(colorRamp(c(\"seagreen\", \"limegreen\", \"yellow3\", \"orange\", \"red\"))(x), maxColorValue = 255)", "\n", sep = "")
        cat("gini_min <- 0", "\n", sep = "")
        cat("gini_max <- 1", "\n", sep = "")
        cat("min_row <- ifelse(nrow(df) > 10, 10, nrow(df))", "\n", sep = "")
        cat("qc_type_exists <- \"screen\" %in% df$`QC_Type`", "\n", sep = "")
        cat("qa_pass_sr <- function(lr_value) {ifelse(lr_value < 30, \"\\u274c\", ifelse(lr_value >= 30 & lr_value < 40, \"\\u2705\\u2757\", \"\\u2705\"))}", "\n", sep = "")
        cat("qa_pass_pl <- function(pass_value) { ifelse(pass_value, \"\\u2705\", \"\\u274c\") }", "\n", sep = "")
        cat("if (!qc_type_exists) df$`% Library Reads` <- qa_pass_pl(df$`% Library Reads`) else df <- df %>% mutate(\"% Library Reads\" = qa_pass_sr(df$QCPass_Library_Per))", "\n", sep = "")
        cat("reactable(df, highlight = TRUE, bordered = TRUE, striped = TRUE, compact = TRUE, wrap = TRUE,", "\n", sep = "")
        cat("          filterable = TRUE, minRows = min_row, defaultColDef = colDef(minWidth = 100, align = \"left\"),", "\n", sep = "")
        cat("          theme = reactableTheme(style = list(fontFamily = \"-apple-system\", fontSize = \"0.85em\")),", "\n", sep = "")
        cat("          columns = list(\"Sample Exon\" = colDef(minWidth = 200),", "\n", sep = "")
        cat("                         \"Group\" = colDef(minWidth = 150),", "\n", sep = "")
        cat("                         \"Sample\" = colDef(minWidth = 150),", "\n", sep = "")
        cat("                         \"Sample Info\" = colDef(minWidth = 150),", "\n", sep = "")
        cat("                         \"Gini Coefficient\" = colDef(style = function(value) {", "\n", sep = "")
        cat("                                                                   normalized <- (value - gini_min) / (gini_max - gini_min)", "\n", sep = "")
        cat("                                                                   list(color = col_pal(normalized))}),", "\n", sep = "")
        cat("                         \"Total Reads\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" }),", "\n", sep = "")
        cat("                         \"% Missing Variants\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" }),", "\n", sep = "")
        cat("                         \"Accepted Reads\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" }),", "\n", sep = "")
        cat("                         \"% Mapping Reads\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" }),", "\n", sep = "")
        cat("                         \"% Reference Reads\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" }),", "\n", sep = "")
        cat("                         \"% Library Reads\" = colDef(filterable = FALSE),", "\n", sep = "")
        cat("                         \"Library Coverage\" = colDef(cell = function(value) { if (value) \"\\u2705\" else \"\\u274c\" }),", "\n", sep = "")
        cat("                         \"QCPass_Library_Per\" = colDef(show = FALSE),", "\n", sep = "")
        cat("                         \"QC_Type\" = colDef(show = FALSE)))", "\n", sep = "")
        cat("```", "\n", sep = "")
        cat("<br>", "\n", sep = "")
        cat("\n", sep = "")

        cat("---", "\n", sep = "")
        cat("\n", sep = "")

        cat("## 4. Methods and Glossary", "\n", sep = "")
        cat("### 4.1. Methods", "\n", sep = "")
        cat("#### 4.1.1. Methods of generating accepted reads (refer to 2.2.3)", "\n", sep = "")
        cat("1) Apply 1D Kmean clustering on each sequence (variant sequence) using log2 read count. Low read count sequences are removed in this step", "\n", sep = "")
        cat("2) A valid sequence must have at least 5 count in at least 25% of the samples in an experiment.", "\n", sep = "")
        cat("\n", sep = "")

        if (qc_type == "screen") {
            cat("#### 4.1.2. Methods of DESeq2 calculation (refer to 2.3.3 and 2.3.4)", "\n", sep = "")
            cat("1) using the total number of accepted reads to calculate the size factor which is applied in DESeq2 normalisation", "\n", sep = "")
            cat("2) run DESeq2 for each consequence", "\n", sep = "")
            cat("3) select synonymous variants and intronic variants as the control, then calculate the median log2 fold change of the control variants", "\n", sep = "")
            cat("4) re-calculate log2 fold change of other consequences by deducting the median log2 fold change of the control variants", "\n", sep = "")
            cat("5) re-calculate p value and adjusted p value", "\n", sep = "")
            cat("\n", sep = "")

            cat("### 4.2. Glossary", "\n", sep = "")
            cat("#### 4.2.1. Glossary of DESeq2 calculation (refer to 2.3.3 and 2.3.4)", "\n", sep = "")
            cat("| name | description |", "\n", sep = "")
            cat("| ----:|:---------------- |", "\n", sep = "")
            cat("| log2FoldChange | This is the initial log2FoldChange from DESeq2 using all the accepted reads |", "\n", sep = "")
            cat("| lfcSE | This is the initial lfcSE (log2FoldChange Standard Error) from DESeq2 using all the accepted reads |", "\n", sep = "")
            cat("| padj | This is the initial corrected p-value from DESeq2 using all the accepted reads |", "\n", sep = "")
            cat("| median control value | This is the median log2 fold change of the control variants (synonymous variants and intronic variants) |", "\n", sep = "")
            cat("| adj_log2FoldChange | This is the adjusted log2FoldChange calculated by deducting the median control value |", "\n", sep = "")
            cat("| adj_score | This is the adjusted score that is calculated from the adj_log2FoldChange divided by the lfcSE |", "\n", sep = "")
            cat("| adj_pval | This is the adjusted p-value derived from adj_score |", "\n", sep = "")
            cat("| adj_fdr | This is the adjusted FDR derived from adj_pval |", "\n", sep = "")
            cat("| stat | This indicates an enriched or a depleted status. adj_fdr < 0.05 & adj_log2FoldChange > 0 is enriched, adj_fdr < 0.05 & adj_log2FoldChange < 0 is depleted |", "\n", sep = "")
        }

        sink()

        rmarkdown::render(paste0(qc_dir, "/MAVEQC_report.Rmd"), clean = TRUE, quiet = TRUE)
        invisible(file.remove(paste0(qc_dir, "/MAVEQC_report.Rmd")))
}